Use Release targets for pre-releases (fixes #90)

mani-monaj opened the pull request:
For pre-release jobs, overwrite `build.targets` feteched from Source build config file with targets fetched from corresponding Release config file. This fixes invalid OS code lookups for pre-release jobs (reported in #90).

tfoote commented:
This is a neat way to mutate the build_file. However I'd prefer it to be explicitly passed instead of mutating the build_file object, there may be other uses of the build_file targets in the future which would want the original data from that field and might incorrectly assume the data is the original. I don't think passing the additional data will be too hard, and it will make the maintainability much better.

mani-monaj commented:
Sounds fair. I'll work on it again later this week and will update this PR.

mani-monaj commented:
@tfoote I implemented your suggestion and updated this PR (it's a single squashed commit).

tfoote commented on ros_buildfarm/devel_job.py:
```diff
@@ -198,6 +200,9 @@ def configure_devel_job(
         build_files = get_source_build_files(config, rosdistro_name)
         build_file = build_files[source_build_name]
+    # Overwrite build_file.targets if build_targets is specified
+    if not build_targets is None:
```
A slight style note: PEP8 calls for 'is not None' it's functionally the same but cleaner to read.

tfoote commented:
This looks good to me. One stylistic comment above.

mani-monaj commented:
Thanks @tfoote. I fixed the stylistic issue and updated the PR.

tfoote commented:
lgtm, can someone else review?

dirk-thomas commented on scripts/prerelease/generate_prerelease_script.py:
```diff
@@ -77,6 +78,12 @@ def main(argv=sys.argv[1:]):
+    # targets defined by Source config. This is to increase the number of
+    # supported targets. New targets are explicitly passed to configure_devel_job().
+    release_build_files = get_release_build_files(config, args.rosdistro_name)
+    release_build_file = release_build_files[args.source_build_name]
```
This won't work well since the `source_build_name` might not be the same. E.g. their is an `arm` release build file with no equivalent source build file.

mani-monaj replied:
You mean `arg.source_build_name` might not be a key in `release_build_files`? If so, is the following a safer alternative:

```python
if release_build_files:
    release_build_file = release_build_files.get(args.source_build_name, None)
```

dirk-thomas replied:
This won;t be sufficient. The index file might define a `default` key for source builds but only a `linux` and `fedora` key for release. There is simply no correlation between those.

mani-monaj replied:
> This won't be sufficient. The index file might define a default key for source builds but only a linux and fedora key for release. There is simply no correlation between those.

In this case and considering the code I recommended `release_build_file` will be set to `None` which is a fallback to the default behavior. Pre-release script will use `build_file = build_files[args.source_build_name]` as the build target (line 79).

dirk-thomas replied:
This won't fix the problem you are aiming to address.

As mentioned above collecting the targets from all release build files seems necessary.

mani-monaj replied:
@dirk-thomas 

`generate_prerelease_script.py` is called with the following parameterts:

```
generate_prerelease_script.py \
  <config_url:https://raw.githubusercontent.com/ros-infrastructure/ros_buildfarm_config/master/index.yaml> \
  <ros_distro:jade> <build_name:default> <os_name:ubuntu> <os_code_name:vivid> <arch:amd64> \
  <pkg_name:ardrone_autonomy> \
  --level 1 \
  --output-dir ./
```

As far as I understood, the script first parses the index file at `config_url` and looks up the `ros_distro` key. Next it looks up `ros_distro/source_builds` key and fetches the YAML config file specified by `ros_disto/source_builds/build_name`. Finally in that config file, it check if `targets/os_name/os_code_name/arch` exists. The problem I am trying to solve here is when the fetched YAML file from `ros_distro/source_builds/build_name` does not define all the targets that exist in the YAML file specified by `ros_distro/release_builds/build_name` key. 

Could you please give some examples of other failure cases? One other failure case that comes to my mind is when the user sets `build_name` to something that is only defined in the release build config file. In case of default index files, `build_name:armhf, arch:armhf` will break the pre-release script.

dirk-thomas replied:
The problem is that there is no correlation between the `build_name` you are using for `source_builds` and `release_builds`. The following `index.yaml` is perfectly valid:

    distributions:
      jade:
        release_builds:
          some_release_build: jade/some-release-build.yaml
          other_release_build: jade/other-release-build.yaml
        source_builds:
          default_source_builds: jade/default-source-build.yaml

Without considering the targets of all release build files you won't be able to use more then the small subset of targets found in the source build file.

mani-monaj replied:
@dirk-thomas Thanks for clarifying the situation. How about something along this line to merge all the targets:

```python
    release_build_files = get_release_build_files(config, args.rosdistro_name)
    release_targets_combined = {}
    release_targets_combined[args.os_name] = {}
    for arch, rel_obj in release_build_files.items():
        for os, t in rel_obj.targets[args.os_name].items():
          if os not in release_targets_combined[args.os_name]:
            release_targets_combined[args.os_name][os] = {}
          release_targets_combined[args.os_name][os].update(t)

  # ...
  # pass release_targets_combined to configure_devel_job()
```

dirk-thomas replied:
Yes, that looks like the best approach. And since the targets defined in the source build file must always be a subset we don't have to consider them.

mani-monaj replied:
Great. I'll refactor/test this, fix minor styling issues and then update this PR.

dirk-thomas commented on ros_buildfarm/devel_job.py:
```diff
@@ -1,6 +1,7 @@
 from __future__ import print_function
 
 import sys
+from copy import deepcopy
```
Please keep imports in alphabetical order following PEP8.

mani-monaj replied:
Sure. I'll fix it.

dirk-thomas commented on scripts/prerelease/generate_prerelease_script.py:
```diff
@@ -24,6 +24,7 @@
 from ros_buildfarm.argument import add_argument_rosdistro_name
 from ros_buildfarm.config import get_index as get_config_index
 from ros_buildfarm.config import get_source_build_files
+from ros_buildfarm.config import get_release_build_files
```
Please keep imports in alphabetical order following PEP8.

mani-monaj replied:
Sure. I'll fix it.

mani-monaj commented:
@dirk-thomas I cleaned up the code to merge all build targets, squashed the changes and updated this PR. It's working fine on my machine, but is it possible for you to review it again and/or test it? Is there any test cases that I can run the script against?

dirk-thomas commented on ros_buildfarm/devel_job.py:
```diff
@@ -198,6 +200,9 @@ def configure_devel_job(
         build_file = build_files[source_build_name]
+    # Overwrite build_file.targets if build_targets is specified
+    if build_targets is not None:
+        build_file.targets = deepcopy(build_targets)
```
I don't think the `deepcopy` is necessary here. The caller should not modify the argument after passing it anyway.

mani-monaj replied:
You are right. I fixed it.

dirk-thomas commented on scripts/prerelease/generate_prerelease_script.py:
```diff
@@ -78,6 +79,22 @@ def main(argv=sys.argv[1:]):
+    # Targets defined by source config file is a subset of targets
+    # defined by release config files. To increase the number of supported
+    # pre-release targets, we combine all targets defined by all release
+    # config files and use that when configuring the devel job.
```
Can you please update the comment to use the term `build file` instead of `config file` since that is the more precise type of these yaml files.

mani-monaj replied:
Sure. fxed.

dirk-thomas commented:
Thank you for updating the PR. I just tried the patch and it works well. I just made two minor comments which should be addressed before the merge.

mani-monaj commented:
@dirk-thomas You are welcome. I applied your fixes and updated the PR.
